Set up database connection and create tables

rvl opened the pull request:
Relates to issue #143 and #154 

# Overview

- [x] Added partial `newDBLayer` function which sets up the db connection and runs migrations.
- [x] Added a small test case to check that the connection works and tables exist.
- [x] Add database queries for wallet metadata.

rvl committed:
DB.Sqlite: Fix type of WalletState column

rvl committed:
Add getPending :: Wallet s t -> Set Tx

rvl committed:
Add PersistField instance for Address

paweljakubas commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -46,6 +46,7 @@ library
     , generic-lens
     , http-api-data
     , http-media
+    , lens
```
we use `generic-lens` already in codebase. If I remember well, it was decided to use this lens library. probably it is good idea in the last commit to comply with the rest of the code (if possible)

paweljakubas replied:
done

paweljakubas commented on lib/core/cardano-wallet-core.cabal:
```diff
@@ -59,6 +60,13 @@ library
+    , persistent-sqlite
+    , conduit
+    , monad-logger
+    , mtl
```
we use `transformers` already. probably it is good idea in the last commit to comply with the rest of the code

paweljakubas replied:
done

rvl committed:
Put back haddocks which got lost in a rebase.

KtorZ commented on lib/core/src/Cardano/Wallet/DB.hs:
```diff
@@ -35,7 +39,7 @@ import Data.Map.Strict
 -- we'll need some extra contraints on the wallet state that allows us to
 -- serialize and unserialize it (e.g. @forall s. (Serialize s) => ...@)
-data DBLayer m s t = DBLayer
+data (IsOurs s, NFData s, Show s, TxId t) => DBLayer m s t = DBLayer
```
`Show s` & `IsOurs s` :thinking: ?

rvl replied:
I have cut it.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -46,7 +128,7 @@ share
     walTableId                 W.WalletId     sql=wallet_id
     walTableName               Text           sql=name
-    walTablePassphraseLastUpdatedAt  UTCTime  sql=passphrase_last_updated_at
+    walTablePassphraseLastUpdatedAt  UTCTime Maybe  sql=passphrase_last_updated_at
```
PTSD triggered :grimacing: ...

rvl replied:
I don't understand what you mean. Should something be changed?

KtorZ replied:
Just the alignment that is off because of the long name. It good for the eyes, but nothing really important here.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1,43 +1,125 @@
 share
-    [ mkPersist sqlSettings { mpsPrefixFields = False }
+    [ mkPersist sqlSettings'
+    , mkDeleteCascade sqlSettings'
```
:pray: Hell yeah. Finally.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -94,7 +177,7 @@ TxIn
 
     Primary txInputTableTxId txInputTableSourceTxId txInputTableSourceIndex
-    Foreign TxMeta fk_tx_meta_tx_in txInputTableTxId txInputTableWalletId
+    Foreign TxMeta fk_tx_meta_tx_in txInputTableTxId txInputTableWalletId -- fixme: remove
```
Why `remove` ? Could the comment give a bit more details about what's wrong, and `how` this should or could be fixed ?

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+runQuery
+    :: SqlBackend
+    -> SqlPersistM a
+    -> ExceptT Sqlite.SqliteException IO a
```
Question: do we know what happens in this `IO` here? Is there any way to lower the constraint down at this level? My concern here is that `runQuery` could be throwing in the wild and bypass the actual `ExceptT` :thinking: ? 

What if the connection is gone, do we get an `SqliteException`, or does it throw? etc .. Is it worth adding some `catch` or `try` and move some exceptions at the value level ?

rvl replied:
I removed this function, but this is a good point.

Any exception that we "expect" (e.g. db constraint), or can recover from, should be put into the DBLayer return types.

For all other exceptions (e.g. connection gone - not sure whether that's even possible in sqlite) then it's fine to exit the wallet with an error message.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+newDBLayer
+    :: forall s t. (IsOurs s, PersistState s, NFData s, Show s, W.TxId t)
+    => Maybe FilePath
+       -- ^ Database file location, or Nothing for in-memory database
```
What about a type isomorphic to `Maybe FilePath` that would convey this idea:

```hs
data Storage = InMemory | InFile FilePath
```

?

rvl replied:
I don't think it's necessary.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+        , putCheckpoint = \(PrimaryKey wid) cp ->
+            ExceptT $ unsafeRunQuery conn $ Right <$> do
+                deleteCheckpoints wid (Proxy :: Proxy s) -- clear out all checkpoints
+                deleteLooseTransactions wid -- clear transactions
```
This is in case we would have some data inconsistency in the database? Is it worth checking the `count` of those queries and, output a warning if anything is beyond `0`?

rvl replied:
This code has been cut and will reappear in a later PR.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+        -----------------------------------------------------------------------}
+
+        , putWalletMeta = \(PrimaryKey wid) meta ->
+            ExceptT $ unsafeRunQuery conn $
```
Why are all the queries running as `unsafeRunQuery` :/ ?

rvl replied:
I made the function safe by renaming it.

KtorZ replied:
> I made the function safe by renaming it.

Hmmm ? :thinking:

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+        }
+
+----------------------------------------------------------------------------
+-- Conversion between Persistent table types and wallet types
```
Seeing conversions below, what about prefixing all db types by `DB` or `SQL`, like `SQLWallet` `SQLPrivateKey` etc ... So that it's obvious where types belong ... (As we do for API types).

rvl replied:
It will be shorter and neater without name prefixes. All these table and column types are local to the Sqlite module. The actual wallet types are qualified with W.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+           | (W.TxIn input ix, W.TxOut addr coin) <- utxoMap ]
+    utxoMap = Map.assocs (W.getUTxO (W.totalUTxO wal))
+
+dist :: (a -> b -> c) -> (a, [b]) -> [c]
```
For small utility functions like that, I find it very useful to put a small doc comment containing an example. It removes some unnecessary brain-load when reading it:

```
-- | Distribute `a` accross many `b`s using the given function.

>>> dist TxOut (addr, [Coin 1, Coin 42, Coin 14])
[TxOut addr (Coin 1), TxOut addr (Coin 42), TxOut addr (Coin 14)]
```

rvl replied:
OK this code has been cut for now but I will address the request for documentation in the next PR.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+    -> s
+    -> W.Wallet s t
+checkpointFromEntity (Checkpoint _ tip) utxo ins outs s =
+    W.Wallet utxo' pending tip s
```
I'd be keen to call this `UnsafeWallet`, or put the constructor in a separate `.Internal` module and add a rule in hlint to make sure that this is only used in this particular file. Exposing the constructor will lead to abuse and since most of our guarantees regarding the wallet state relies on the fact that only "applyBlock" is used to modify the state of the wallet, it is rather important to enforce we preserve that.

rvl replied:
This code has been deleted and will reappear in a later PR.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -1,43 +1,125 @@
 
+import qualified Cardano.Wallet.Primitive.AddressDiscovery as W
+import qualified Cardano.Wallet.Primitive.Model as W
 import qualified Cardano.Wallet.Primitive.Types as W
```
The `W.` across the file feels very noisy to me :disappointed: ... I am not sure about this style.

rvl replied:
It is very reasonable to have qualified name prefixing in a module where both SQL table types and application types are present.

paweljakubas replied:
in previous codebase we had `Kernel` prefix

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+    insertState :: MonadIO m => (W.WalletId, W.SlotId) -> s -> ReaderT SqlBackend m ()
+    selectState :: MonadIO m => (W.WalletId, W.SlotId) -> ReaderT SqlBackend m (Maybe s)
+    deleteState :: MonadIO m => Proxy s -> W.WalletId -> ReaderT SqlBackend m ()
+    addressScheme :: Proxy s -> AddressScheme
```
I am not sure to understand the rationale behind `insertState` and `selectState` ? What are their use? Can we add comments please :pray:

rvl replied:
This code has been cut and will appear in a second PR with comments.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +188,589 @@ TxOut
+    }
+    where getAmount (Quantity n) = n
+
+-- note: TxOut records must already be sorted by index
```
Sounds like a good property-test to have, or an invariant in the code. Either way, the note isn't sufficient here.

rvl replied:
This code has been cut and will reappear in a subsequent PR.

KtorZ reviewed (commented):
Looks good overall. Although the `Sqlite` file lacks of documentation IMO. Don't forget that we are the first consumers of that code and that documentation. Be verbose on details, give concrete examples, make it easy to consume the module :pray:

rvl committed:
Sqlite: Add on delete cascade to schema

rvl committed:
Sqlite: Fix instance PersistField TxStatus

It had an infinite loop.

rvl committed:
Sqlite: fix the schema

rvl committed:
tests: Add simple Sqlite DB layer test

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery.hs:
```diff
@@ -42,9 +42,10 @@ module Cardano.Wallet.Primitive.AddressDiscovery
     , accountPubKey
     , mkAddressPool
     , lookupAddress
+    , indexedAddresses -- fixme: internals exposed for sql
```
NOTE: `addresses` is already exposed and does exactly what the SQL engine needs I believe.

rvl committed:
Sqlite: Add database connection and wallet metadata queries

akegalj commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +153,221 @@ TxOut
+        { createWallet = \(PrimaryKey wid) _cp meta ->
+            ExceptT $ unsafeRunQuery conn $
+            selectWallet wid >>= \case
+                Just _ ->
```
what would happen if we didn't make this check? Would `insert_` fail?

paweljakubas replied:
if we do not check this then we cannot handle `ErrWalletAlreadyExists` - there are tests that are checking this and my believe is that we do not want to do it

rvl replied:
There would be an exception from SQLite. Actually I think it will be better to handle this exception than do select + insert. Fixed.

akegalj commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +153,221 @@ TxOut
+metadataFromEntity wal = W.WalletMetadata
+    { name = W.WalletName (walTableName wal)
+    , passphraseInfo = case walTablePassphraseLastUpdatedAt wal of
+            Just time -> Just $ W.WalletPassphraseInfo time
```
why not 
```
passphraseInfo = W.WalletPassphraseInfo <$> walTablePassphraseLastUpdatedAt wal
```

rvl replied:
Yep, fixed.

akegalj commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +153,221 @@ TxOut
+    [ WalTableName =. meta ^. #name . coerce
+    , WalTablePassphraseLastUpdatedAt =. case meta ^. #passphraseInfo of
+            Nothing -> Nothing
+            Just (W.WalletPassphraseInfo passInfo) -> Just passInfo
```
why not:
```
WalTablePassphraseLastUpdatedAt =.  W.lastUpdateAt <$> meta ^. #passphraseInfo

akegalj commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -105,49 +153,221 @@ TxOut
+    , walTableName = meta ^. #name . coerce
+    , walTablePassphraseLastUpdatedAt = case meta ^. #passphraseInfo of
+            Nothing -> Nothing
+            Just (W.WalletPassphraseInfo passInfo) -> Just passInfo
```
https://github.com/input-output-hk/cardano-wallet/pull/247/files#r284836455

rvl committed:
fixup! Sqlite: Add database connection and wallet metadata queries

rvl committed:
fixup! tests: Add simple Sqlite DB layer test

rvl committed:
fixup! Sqlite: fix the schema

rvl committed:
Sqlite: use exceptions for handling ErrWalletAlreadyExists

rvl committed:
Sqlite: Some testing of exception handling

rvl committed:
Sqlite: add a couple source comments

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -103,51 +158,219 @@ TxIn
+       -- ^ Database file location, or Nothing for in-memory database
+    -> IO (DBLayer IO s t)
+newDBLayer fp = do
+    conn <- createSqliteBackend fp logStderr
```
wil this `logStderr` enable all logs (CREATE_TABLE, INSERT, ...) as a default or errors only? I would give my votes for the second option

paweljakubas replied:
indeed, it introduces logs like : 
```Cardano.Wallet.DB.Sqlite
  Wallet table
Migrating: CREATE TABLE "wallet"("wallet_id" VARCHAR NOT NULL,"name" VARCHAR NOT NULL,"passphrase_last_updated_at" TIMESTAMP NULL,"status" INTEGER NOT NULL,"delegation" VARCHAR NULL, PRIMARY KEY ("wallet_id"))
CREATE TABLE "wallet"("wallet_id" VARCHAR NOT NULL,"name" VARCHAR NOT NULL,"passphrase_last_updated_at" TIMESTAMP NULL,"status" INTEGER NOT NULL,"delegation" VARCHAR NULL, PRIMARY KEY ("wallet_id")); []
```

paweljakubas replied:
added more customizable logging function (one can choose LogLevel)

paweljakubas commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -103,51 +158,219 @@ TxIn
+        W.lastUpdatedAt <$> meta ^. #passphraseInfo
+    , walTableStatus = meta ^. #status
+    , walTableDelegation = delegationToText $ meta ^. #delegation
+    , walTableAddressScheme = Sequential
```
This probably has to be addressed in another PR. Either removed or detected whether we are Random or Sequential

rvl replied:
Yes there is not yet reading/writing of the wallet state.
However we may even remove this field because the address scheme is fixed by a type parameter of DBLayer.

paweljakubas replied:
exactly. This is in line what @KtorZ wrote on slack yesterday:
```
Then, we'll have one

- external seq wallet layer + seq db layer
- fully owned  seq wallet + seq db layer
- fully owned rnd wallet + rnd db layer
```
Address scheme is set for each db instance. Hence, `walTableAddressScheme` is probably superflouous

paweljakubas replied:
removed address scheme

paweljakubas reviewed (approved):
LGTM

rvl committed:
Sqlite: add a test of cascading delete

This will show whether cascading deletes work under sqlite -- once the
functions to write other tables are implemented.

paweljakubas committed:
Remove superfluous address scheme from db

paweljakubas committed:
use beforeAll and beforeWith combination in DBSpec

paweljakubas committed:
Control more logging from db
